Userland/Applications/HexEditor/EditAnnotationDialog.cpp

```diff
@@ -0,0 +1,109 @@
+ */
+
+#include "EditAnnotationDialog.h"
+#include <LibGUI/MessageBox.h>
+
+static Gfx::Color s_most_recent_color { Color::from_argb(0xfffce94f) };
+
+GUI::Dialog::ExecResult EditAnnotationDialog::show_create_dialog(GUI::Window* parent_window, HexDocument& document, Selection selection)
+{
+    auto dialog_or_error = EditAnnotationDialog::try_create(parent_window, document, selection);
+    if (dialog_or_error.is_error()) {
+        GUI::MessageBox::show(parent_window, MUST(String::formatted("{}", dialog_or_error.error())), "Error while opening Create Annotation dialog"sv, GUI::MessageBox::Type::Error);
+        return ExecResult::Aborted;
+    }
+
+    auto dialog = dialog_or_error.release_value();
+    return dialog->exec();
+}
+
+GUI::Dialog::ExecResult EditAnnotationDialog::show_edit_dialog(GUI::Window* parent_window, HexDocument& document, Annotation& annotation)
+{
+    auto dialog_or_error = EditAnnotationDialog::try_create(parent_window, document, &annotation);
+    if (dialog_or_error.is_error()) {
+        GUI::MessageBox::show(parent_window, MUST(String::formatted("{}", dialog_or_error.error())), "Error while opening Edit Annotation dialog"sv, GUI::MessageBox::Type::Error);
+        return ExecResult::Aborted;
+    }
+
+    auto dialog = dialog_or_error.release_value();
+    return dialog->exec();
+}
+
+ErrorOr<NonnullRefPtr<EditAnnotationDialog>> EditAnnotationDialog::try_create(GUI::Window* parent_window, HexDocument& hex_document, Variant<Annotation*, Selection> selection_or_annotation)
+{
+    auto widget = TRY(HexEditor::EditAnnotationWidget::try_create());
+    return adopt_nonnull_ref_or_enomem(new (nothrow) EditAnnotationDialog(parent_window, move(widget), hex_document, move(selection_or_annotation)));
+}
+
+EditAnnotationDialog::EditAnnotationDialog(GUI::Window* parent_window, NonnullRefPtr<HexEditor::EditAnnotationWidget> widget, HexDocument& hex_document, Variant<Annotation*, Selection> selection_or_annotation)
+    : GUI::Dialog(parent_window)
+    , m_document(hex_document.make_weak_ptr())
+{
+    resize(260, 140);
+    set_resizable(false);
+    set_main_widget(widget);
+
+    m_start_offset = find_descendant_of_type_named<GUI::NumericInput>("start_offset");
+    m_end_offset = find_descendant_of_type_named<GUI::NumericInput>("end_offset");
+    m_background_color = find_descendant_of_type_named<GUI::ColorInput>("background_color");
+    m_save_button = find_descendant_of_type_named<GUI::DialogButton>("save_button");
+    m_cancel_button = find_descendant_of_type_named<GUI::DialogButton>("cancel_button");
+
+    // FIXME: This could be specified in GML, but the GML doesn't like property setters that aren't `set_FOO()`.
+    m_background_color->set_color_has_alpha_channel(false);
+
+    // NOTE: The NumericInput stores an i64, so not all size_t values can fit. But I don't think we'll be
+    //       hex-editing files larger than 9000 petabytes for the foreseeable future!
+    VERIFY(hex_document.size() <= INT64_MAX);
```

Comment:
Totally a nitpick, but in the light of #21995, can we do `NumericLimits<i64>::max()` here and further down the file?

Reply:
Sure! I couldn't remember how to do that properly when I wrote this.